builtin/logical/totp/path_keys.go

```diff
@@ -0,0 +1,400 @@
+package totp
+
+import (
+	"bytes"
+	"encoding/base32"
+	"encoding/base64"
+	"fmt"
+	"image/png"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+	otplib "github.com/pquerna/otp"
+	totplib "github.com/pquerna/otp/totp"
+)
+
+func pathListKeys(b *backend) *framework.Path {
+	return &framework.Path{
+		Pattern: "keys/?$",
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.ListOperation: b.pathKeyList,
+		},
+
+		HelpSynopsis:    pathKeyHelpSyn,
+		HelpDescription: pathKeyHelpDesc,
+	}
+}
+
+func pathKeys(b *backend) *framework.Path {
+	return &framework.Path{
+		Pattern: "keys/" + framework.GenericNameRegex("name"),
+		Fields: map[string]*framework.FieldSchema{
+			"name": {
+				Type:        framework.TypeString,
+				Description: "Name of the key.",
+			},
+
+			"generate": {
+				Type:        framework.TypeBool,
+				Default:     false,
+				Description: "Determines if a key is generated by vault or another service.",
+			},
+
+			"key_size": {
+				Type:        framework.TypeInt,
+				Default:     20,
+				Description: "Determines the size in bytes of the generated key. Only used if generate is true.",
+			},
+
+			"key": {
+				Type:        framework.TypeString,
+				Description: "The shared master key used to generate a TOTP token. Only used if generate is false.",
+			},
+
+			"issuer": {
+				Type:        framework.TypeString,
+				Description: `The name of the key's issuing organization. Required if generate is true and url is not passed.`,
+			},
+
+			"account_name": {
+				Type:        framework.TypeString,
+				Description: `The name of the account associated with the key. Required if generate is true and url is not passed.`,
+			},
+
+			"period": {
+				Type:        framework.TypeDurationSecond,
+				Default:     30,
+				Description: `The length of time used to generate a counter for the TOTP token calculation.`,
+			},
+
+			"algorithm": {
+				Type:        framework.TypeString,
+				Default:     "SHA1",
+				Description: `The hashing algorithm used to generate the TOTP token. Options include SHA1, SHA256 and SHA512.`,
+			},
+
+			"digits": {
+				Type:        framework.TypeInt,
+				Default:     6,
+				Description: `The number of digits in the generated TOTP token. This value can either be 6 or 8.`,
+			},
+
+			"skew": {
+				Type:        framework.TypeInt,
+				Default:     0,
```

Comment:
I think most implementations use a value of 1 for this, e.g. tolerate a 30-second window on either side. Is there a reason to keep it so tight rather than defaulting to the "norm"?